taichi/transforms/binary_op_simplify.cpp

```diff
@@ -0,0 +1,94 @@
+#include "taichi/ir/ir.h"
+#include "taichi/ir/transforms.h"
+#include "taichi/ir/visitors.h"
+#include "taichi/program/program.h"
+
+TLANG_NAMESPACE_BEGIN
+
+class BinaryOpSimp : public BasicStmtVisitor {
+ public:
+  using BasicStmtVisitor::visit;
+  bool fast_math;
+  DelayedIRModifier modifier;
+
+  explicit BinaryOpSimp(bool fast_math_)
+      : BasicStmtVisitor(), fast_math(fast_math_) {
+  }
+
+  void visit(BinaryOpStmt *stmt) override {
+    // swap lhs and rhs if lhs is a const and op is commutative
+    auto const_lhs = stmt->lhs->cast<ConstStmt>();
+    if (const_lhs && is_commutative(stmt->op_type)) {
+      auto rhs_stmt = stmt->rhs;
+      stmt->lhs = rhs_stmt;
+      stmt->rhs = const_lhs;
+    }
+    if (!fast_math) {
+      return;
+    }
+    auto binary_lhs = stmt->lhs->cast<BinaryOpStmt>();
+    auto const_rhs = stmt->rhs->cast<ConstStmt>();
+    if (!binary_lhs || !const_rhs) {
+      return;
+    }
+    auto const_lhs_rhs = binary_lhs->rhs->cast<ConstStmt>();
+    if (!const_lhs_rhs) {
+      return;
+    }
+    // original:
+    // stmt = (a op1 b) op2 c
+    // rearrange to:
+    // stmt = a op1 (b op2 c)
+    if (can_rearrange_associative(binary_lhs->op_type, stmt->op_type)) {
+      auto bin_op =
+          Stmt::make<BinaryOpStmt>(stmt->op_type, const_lhs_rhs, const_rhs);
+      bin_op->ret_type.data_type = stmt->ret_type.data_type;
+      stmt->lhs = binary_lhs->lhs;
+      stmt->op_type = binary_lhs->op_type;
+      stmt->rhs = bin_op.get();
+      stmt->parent->insert_before(stmt, std::move(bin_op));
```

Comment:
Let's use `modifier->insert_before`. `stmt->parent->insert_before` modifies IR directly and may invalidate the `std::vector::iterator`s in `BasicStmtVisitor`.

BTW `stmt->lhs` can be `modifier->erase`d.

Reply:
Oh BTW we don't want to modify non-operand fields (i.e. non-`Stmt *` fields) of statements directly -- currently, some parts of code like `same_statements` presume that statements' fields are unchanged after the constructor. Let's simply make a new statement, replace `stmt` with it, insert it before `stmt` and (delayed) erase `stmt`.

Reply:
I am not sure a local swap can ensure `stmt->lhs` can be safely removed, consider this made-up case:
```
$4 = add $2 $3
$5 = add $4 $2
$6 = add $4 $3
```
if we erase $4 when we rearrange $5, $6 will lose its operand.

Reply:
Good catch. We can't erase `stmt->lhs` here.